marimo/_server/server_handler.py

```diff
@@ -0,0 +1,269 @@
+                LOGGER.debug("Reconnecting session %s", session_id)
+                self.reconnect_session(session)
+            else:
+                # No clients are connected, and we haven't seen this session id
+                # before. Create a session.
+                #
+                # If the client refreshed their page, there will be one
+                # existing session with a closed socket for a different session
+                # id; that's why we call `close_all_sessions`.
+                mgr.close_all_sessions()
+                mgr.create_session(
+                    session_id=session_id,
+                    session_handler=TornadoWebsocketSessionHandler(self),
+                )
+                self.status = ConnectionState.OPEN
+                self.write_kernel_ready()
+        elif session is not None:
+            LOGGER.debug("Reconnecting session %s", session_id)
+            # Cancel previous close handle
+            self.cancel_close_handle = None
+            self.reconnect_session(session)
+        else:
+            mgr.create_session(
+                session_id=session_id,
+                session_handler=TornadoWebsocketSessionHandler(self),
+            )
+            self.status = ConnectionState.OPEN
+            # Let the frontend know it can instantiate the app.
+            self.write_kernel_ready()
+
+    def on_close(self) -> None:
+        """
+        When the websocket is closed, we wait TTL_SECONDS before closing the
+        session. This is to prevent the session from being closed if the
+        during an intermittent network issue.
+        """
+        LOGGER.debug("websocket connection for %s closed", self.session_id)
+        self.status = ConnectionState.CLOSED
+        mgr = get_manager()
+        if mgr.mode == SessionMode.RUN:
+
+            def _close() -> None:
+                if self.status != ConnectionState.OPEN:
+                    LOGGER.debug(
+                        "Closing session %s (TTL EXPIRED)", self.session_id
+                    )
+                    mgr.close_session(self.session_id)
+
+            session = mgr.get_session(self.session_id)
+            cancellation_handle = tornado.ioloop.IOLoop.current().call_later(
+                Session.TTL_SECONDS, _close
+            )
+            if session is not None:
+                self.cancel_close_handle = cancellation_handle
+
+
+class TornadoWebsocketSessionHandler(SessionHandler):
+    """
+    Implements a session handler for tornado websockets
+    """
+
+    mode: Optional[SessionMode] = None
+
+    def __init__(self, websocket: IOSocketHandler):
+        self.websocket = websocket
+
+    def on_start(
+        self,
+        mode: SessionMode,
+        connection: Connection,
+        check_alive: Callable[[], None],
+    ) -> None:
+        self.mode = mode
+        if mode == SessionMode.RUN:
+            loop = tornado.ioloop.IOLoop.current()
+            loop.remove_handler(connection.fileno())
```

Comment:
Isn't this clean up logic, which happens in `on_stop`? Why is this here?

Reply:
incorrectly copied from the previous code. fixed